src/main/java/com/faforever/moderatorclient/ui/LoginController.java

```diff
@@ -35,32 +41,53 @@ public Pane getRoot() {
 
     @FXML
     public void initialize() {
-        errorMessageLabel.managedProperty().bind(errorMessageLabel.visibleProperty());
-        errorMessageLabel.setVisible(false);
-
         applicationProperties.getEnvironments().forEach(
                 (key, environmentProperties) -> environmentComboBox.getItems().add(key)
         );
 
+        environmentComboBox.setOnAction((event) -> loginWebView.getEngine().load(getHydraUrl()));
+
         environmentComboBox.getSelectionModel().select(0);
-    }
 
-    public void onLoginClicked() {
-        EnvironmentProperties environmentProperties = applicationProperties.getEnvironments()
-                .get(environmentComboBox.getSelectionModel().getSelectedItem());
-        fafApiCommunicationService.initialize(environmentProperties);
+        loginWebView.getEngine().load(getHydraUrl());
+
+        loginWebView.getEngine().locationProperty().addListener((observable, oldValue, newValue) -> {
+            int codeIndex = newValue.indexOf("code=");
+            if (codeIndex >= 0) {
+                int codeEnd = newValue.indexOf("&", codeIndex);
+                String code = newValue.substring(codeIndex + 5, codeEnd);
+                int stateIndex = newValue.indexOf("state=");
+                int stateEnd = newValue.indexOf("&", stateIndex);
+                String reportedState;
+                if (stateEnd > 0) {
+                    reportedState = newValue.substring(stateIndex + 6, stateEnd);
+                } else {
+                    reportedState = newValue.substring(stateIndex + 6);
+                }
 
```

Comment:
Why not parse the arguments using 
`URLEncodedUtils.parse(new URI("https://stackoverflow.com/questions/13592236/parse-a-uri-string-into-name-value-collection?x=1&y=2"), StandardCharsets.UTF_8)  ====>  [x=1, y=2]`

Reply:
Probably because I did not know about it